design/src/rfcs/rfc0034_generating_reusable_code.md

```diff
@@ -0,0 +1,232 @@
+# RFC: Principles for Generating Reusable Code
+
+> Status: Accepted
+
+One might characterize `smithy-rs` as a way of converting [Smithy shapes](https://smithy.io/2.0/spec/model.html#shapes) into Rust types. The generated Rust types are the customer facing interface and, for this reason, they should accurately and predictably reflect the properties of the shapes they represent. One property that Smithy shapes enjoy is reusability. This RFC proposes simple guiding principles to ensure that reusability of Rust types reflects the reusability of Smithy shapes.
+
+This RFC is does _not_ propose specific changes to the code generators API or the client or server Rust API. However the principles, when applied, are intended to enable desirable APIs for both the code generator and the client and server SDK.
+
+## Terminology
+
+- **Compatible types**: Two types are said to be "compatible", if they could in principle be collapsed into a single type. That is, if they have the identical type definition and non-conflicting methods/trait implementations.
+- **Abstract Syntax Tree (AST)**: A tree representing the _syntactic_ structure of the source code of a formal language.
+- **Abstract Semantic Graph (ASG)**: A graph representing the _semantic_ structure of the source code of a formal language. In our context, one can think of an ASG as an AST with additional edges introduced between nodes by the presence of identifiers. For example, a [Smithy Shape](https://smithy.io/2.0/spec/model.html#shapes), represented by a node in the ASG, would enjoy edges to each of its member shapes.
```

Comment:
Hence I'd remove the AST definition entirely, since it's not referred to in the body of the RFC; it's only used to introduce the ASG concept in a rather confusing way.

Reply:
I included it because I wanted to piggyback off the definition a little and it is mentioned briefly to say that 

> At high-level, the [Smithy library](https://github.com/awslabs/smithy-rs) library provides not only a way to parse the AST defined by the...

But yeah, I agree this has turned out badly. I will try again with concrete references to what it means for Smithy.